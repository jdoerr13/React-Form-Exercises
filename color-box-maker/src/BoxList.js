import React, { useState } from "react";
import { v4 as uuidv4 } from 'uuid';//provides a function to get a random id
import Box from "./Box"
import NewBoxForm from "./NewBoxForm";

//PARENT COMPONENT THAT WILL HOLD THE STATE: OR container component that manages the state of the shopping list and coordinates the addition of new items through the NewItemForm component.

//MANAGES THE OVERALL LIST OF BOXES- ADD & REMOVE AND CHILD COMPONENETS
const BoxList = () => {
//no innitial state?
const INITIAL_STATE = [
  { id: uuidv4(), width: 5, height: 5, backgroundColor: 'red' },
  { id: uuidv4(), width: 7, height: 4, backgroundColor: 'blue' }
];
  const [boxes, setBoxes] = useState(INITIAL_STATE);

  //will no longer just receive name and quantity= newItem is the object from ..formData. therefore it is just a PARAMETER
  const addBox = (boxData) => {
    // Create a new box with a unique ID and the rest of the boxData
    const newBox = { ...boxData, id: uuidv4() };
    setBoxes((boxes) => [...boxes, newBox]);
  };

  const removeBox = (id) => {
    // Filter out the box with the matching ID
    setBoxes((boxes) => boxes.filter((box) => box.id !== id));
  };

  return (
    <div>
      <NewBoxForm addBox={addBox} />
      {boxes.map((box) => (
        <Box
          key={box.id}
          id={box.id}
          width={box.width}
          height={box.height}
          backgroundColor={box.backgroundColor}
          removeBox={() => removeBox(box.id)}
        />
      ))}
    </div>
  );
}

export default BoxList;


