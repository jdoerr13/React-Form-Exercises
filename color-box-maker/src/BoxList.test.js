import React from "react";
import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});

it("should add new item", function () {
  //select various pieces by queryByText & getByLabelText
  const { container, getByText, getByLabelText } = render(<BoxList />);

  const wInput = getByLabelText("Width:");
  const hInput = getByLabelText("Height:");
  const bcInput = getByLabelText("Background Color:");
  const btn = getByText("Add New Box");

  fireEvent.change(wInput, { target: { value: '5' } });
  fireEvent.change(hInput, { target: { value: '5' } });
  fireEvent.change(bcInput, { target: { value: 'green' } });

  // Submit the form
  fireEvent.click(btn);

  const addedBox = [...container.querySelectorAll('div')].find(
    (div) => div.style.width === '5em' && 
             div.style.height === '5em' && 
             div.style.backgroundColor === 'green'
             );
  expect(addedBox).toBeInTheDocument();
})


it("should remove a box", function () {
  const { container, getByText, getByLabelText, getAllByText } = render(<BoxList />);

  const wInput = getByLabelText("Width:");
  const hInput = getByLabelText("Height:");
  const bcInput = getByLabelText("Background Color:");
  const btn = getByText("Add New Box");

  fireEvent.change(wInput, { target: { value: '8' } });
  fireEvent.change(hInput, { target: { value: '5' } });
  fireEvent.change(bcInput, { target: { value: 'green' } });

  // Submit the form
  fireEvent.click(btn);

  const addedBox = [...container.querySelectorAll('div')].find(
    (div) => div.style.width === '8em' && 
             div.style.height === '5em' && 
             div.style.backgroundColor === 'green'
             );
  expect(addedBox).toBeInTheDocument();

   // Step 2: Remove the todo
   const deleteButton = addedBox.querySelector('button'); 
   fireEvent.click(deleteButton);
 
   // Step 3: Verify box is removed
   expect(addedBox).not.toBeInTheDocument();
 });