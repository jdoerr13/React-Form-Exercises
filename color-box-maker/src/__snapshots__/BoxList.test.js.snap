// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <form>
      <div>
        <label
          for="width"
        >
          Width:
        </label>
        <input
          id="width"
          name="width"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="height"
        >
          Height:
        </label>
        <input
          id="height"
          name="height"
          type="text"
          value=""
        />
      </div>
      <div>
        <label
          for="backgroundColor"
        >
          Background Color:
        </label>
        <input
          id="backgroundColor"
          name="backgroundColor"
          type="text"
          value=""
        />
      </div>
      <button
        type="submit"
      >
        Add New Box
      </button>
    </form>
    <div
      style="width: 5em; height: 5em; background-color: red;"
    >
      <button>
        X
      </button>
    </div>
    <div
      style="width: 7em; height: 4em; background-color: blue;"
    >
      <button>
        X
      </button>
    </div>
  </div>
</DocumentFragment>
`;
